src/vector_commitment/vector_commitment.cairo

```diff
@@ -0,0 +1,175 @@
+}
+
+// Decommits a VectorCommitment at multiple indices.
+// Indices must be sorted and unique.
+fn vector_commitment_decommit(
+    commitment: VectorCommitment, queries: Span<VectorQuery>, witness: VectorCommitmentWitness,
+) {
+    let shift = pow(2, commitment.config.height);
+    let shifted_queries = shift_queries(queries, shift, commitment.config.height);
+
+    let expected_commitment = compute_root_from_queries(
+        shifted_queries,
+        0,
+        commitment.config.n_verifier_friendly_commitment_layers,
+        witness.authentications,
+        0
+    );
+
+    assert(expected_commitment == commitment.commitment_hash, 'decommitment failed');
+}
+
+// Verifies a queue of Merkle queries. [queue_head, queue_tail) is a queue, where each element
+// represents a node index (given in a heap-like indexing) and value (either an inner
+// node or a leaf).
+fn compute_root_from_queries(
+    mut queue: Array<VectorQueryWithDepth>,
+    start: u32,
+    n_verifier_friendly_layers: felt252,
+    authentications: Span<felt252>,
+    auth_start: u32
+) -> felt252 {
+    let current: VectorQueryWithDepth = *queue[start];
+
+    if current.index == 1 { // root
+        assert(current.depth == 0, 'root depth must be 0');
+        assert(start + 1 == queue.len(), 'root must be the last element');
+        assert(auth_start == authentications.len(), 'authentications is too long');
+        return current.value;
+    }
+
+    let (parent, bit) = div_rem2(current.index);
+    let is_verifier_friendly = n_verifier_friendly_layers >= current.depth;
+    let hash = if bit == 0 {
+        if start + 1 != queue.len() {
+            let next: VectorQueryWithDepth = *queue[start + 1];
+            if current.index + 1 == next.index {
+                // next is a sibling of current
+                let hash = hash_blake_or_poseidon(current.value, next.value, is_verifier_friendly);
+                queue
+                    .append(
+                        VectorQueryWithDepth {
+                            index: parent, value: hash, depth: current.depth - 1,
+                        }
+                    );
+                return compute_root_from_queries(
+                    queue, start + 2, n_verifier_friendly_layers, authentications, auth_start
+                );
+            }
+        }
+        assert(auth_start != authentications.len(), 'authentications is too short');
+        hash_blake_or_poseidon(current.value, *authentications[auth_start], is_verifier_friendly)
+    } else {
+        assert(auth_start != authentications.len(), 'authentications is too short');
+        hash_blake_or_poseidon(*authentications[auth_start], current.value, is_verifier_friendly)
+    };
+    queue.append(VectorQueryWithDepth { index: parent, value: hash, depth: current.depth - 1, });
+    compute_root_from_queries(
+        queue, start + 1, n_verifier_friendly_layers, authentications, auth_start + 1
+    )
+}
```

Comment:
```suggestion
// Verifies a queue of Merkle queries. [queue_head, queue_tail) is a queue, where each element
// represents a node index (given in a heap-like indexing) and value (either an inner
// node or a leaf).
fn compute_root_from_queries(
    mut queue: Array<VectorQueryWithDepth>,
    n_verifier_friendly_layers: felt252,
    authentications: Span<felt252>,
) -> felt252 {
    let mut queue_i: u32 = 0;
    let mut authentications_i: u32 = 0;

    loop {
        let current: VectorQueryWithDepth = *queue.at(queue_i);
        if current.index == 1 { // root
            assert(current.depth == 0, 'root depth must be 0');
            assert(queue_i + 1 == queue.len(), 'root must be the last element');
            assert(authentications_i == authentications.len(), 'authentications is too long');
            break current.value;
        }

        let (parent, bit) = div_rem2(current.index);
        let is_verifier_friendly = n_verifier_friendly_layers >= current.depth;
        let hash = if bit == 0 {
            if queue_i + 1 != queue.len() {
                let next: VectorQueryWithDepth = *queue.at(queue_i + 1);
                if current.index + 1 == next.index {
                    // next is a sibling of current
                    let hash = hash_blake_or_poseidon(
                        current.value, next.value, is_verifier_friendly
                    );
                    queue
                        .append(
                            VectorQueryWithDepth {
                                index: parent, value: hash, depth: current.depth - 1,
                            }
                        );
                    queue_i += 2;
                    continue;
                }
            }
            assert(authentications_i != authentications.len(), 'authentications is too short');
            hash_blake_or_poseidon(
                current.value, *authentications.at(authentications_i), is_verifier_friendly
            )
        } else {
            assert(authentications_i != authentications.len(), 'authentications is too short');
            hash_blake_or_poseidon(
                *authentications.at(authentications_i), current.value, is_verifier_friendly
            )
        };
        queue.append(VectorQueryWithDepth { index: parent, value: hash, depth: current.depth - 1 });
        queue_i += 1;
        authentications_i += 1;
    }
}
```

Reply:
I tested iterative version previously and it was slower. So do we want it?

Reply:
Yes my results are the same iterative is slower a bit.
I like iterative version but that is only my subjective preference only
Lets wait for @tiagofneto 's opinion

Reply:
Honestly both look good and readable for me, let's go with the faster one (recursive I guess)